### PR TITLE
Apply the latest Beman Standard

### PR DESCRIPTION
**Pull request by neatudarius**
Apply the [Beman Standard](https://github.com/beman-project/beman/blob/main/docs/beman-standard.md) to Beman.Example and more tweaks.

Issues:
- #14 
- #16
- #9
- #8 

Updates related to the Beman Standard:
- [x] Update licence
   - [x]  Apply `LICENSE.APACHE_LLVM`: Switch from MIT to `Apache License v2.0 with LLVM Exceptions`.
   - [x]  Apply `LICENSE.CRITERIA`: already applied
   - [x] Apply `LICENSE.APPROVED`: already applied 
- [x] Apply `FILE.COPYRIGHT`
- [x] Naming
   - [x]  Apply `LIBRARY_NAMES`:
   - [x] Apply `REPOSITORY_NAME`: renamed to `Example` (docs + github)
   - [x] Apply `TOPLEVEL.CMAKE`: already applied (root `CMakeLists.txt`, using Ctest)
   - [x] Apply `TOPLEVEL.LICENSE`: renamed file to `LICENCE`.
   - [x] Apply `TOPLEVEL.README` 
- [x] README
   - [x] Apply `README.TITLE`
   - [x] Apply `README.PURPOSE`
   - [x] Apply `README.IMPLEMENTS`
- [x] CMake
    - [x] Apply `CMAKE.DEFAULT` 
    - [x] Apply `CMAKE.SKIP_TESTS`
    - [x] Apply `CMAKE.AVOID_PASSTHROUGHS`
- [x] Tree
    - [x] Apply `DIRECTORY.INTERFACE_HEADERS`: 
        ```shell
        $ tree include 
        include
        └── Beman
            └── Example
                └── example.hpp
        ```
    - [x] Apply `DIRECTORY.IMPLEMENTATION_HEADERS`: currently N/A
    - [x] Apply `DIRECTORY.SOURCES`:
        ```shell
        $ tree src 
        src
        └── Beman
            └── Example
                ├── CMakeLists.txt
                ├── example.cpp
                └── tests
                    ├── CMakeLists.txt
                    └── example.t.cpp
        
        4 directories, 4 files
        ```

Updates related to docs and usage example:
- [x] Add API example: `beman::example::identity` equivalent of [std::identity](https://eel.is/c++draft/func.identity)
- [x] Add basic unit test and framework integration with Gtest: `src/Beman/Example/tests/example.t.cpp`.
```shell
$ tree src/Beman/Example/tests 
src/Beman/Example/tests
├── CMakeLists.txt
└── example.t.cpp

1 directory, 2 files
```
- [x] Add usage example: `examples/sample_usage.cpp`.
```shell
$ tree examples 
examples
├── CMakeLists.txt
└── sample_usage.cpp

1 directory, 2 files
```
- [ ] Docs
    - [ ] Add build instructions docs.
    - [ ] Document minimum compiler requirements
    - [ ] Add required tools/packags.
    - [ ] Add OS support.
    - [ ] Add deps support.
    - [ ] Add C++ standard support.

Updates related to CI/infrastructure/other missing bits:
- [x] Extend compiler /  standard versions for build and run on CI: {`gcc`, `clang`}` x {`C++98`, `C++03`, `C++11`, `C++14`, `C++17`, `C++20`, 'C++23`, `C++26`}.
- [x] Add CI Tests badge in root README.md.
- [x] Add `.gitignore`
- [x] Add `.github/CODEOWNERS`

### CODE REVIEW

File: .github/CODEOWNERS
Comment (neatudarius) [inline .github/CODEOWNERS:3]: Please use this thread to comment if we want or not to add this files.

If yes, should we added it the standard?

If yes, people don't have pemissions - CC: @camio - `Unknown owner on line 3: make sure @neatudarius exists and has write access to the repository`.

File: LICENSE
Comment (neatudarius) [inline LICENSE:1]: Should we switch to `Apache License v2.0 with LLVM Exceptions`  as per `LICENSE.APACHE_LLVM`?

File: include/Beman/Example/example.hpp
Comment (neatudarius) [inline include/Beman/Example/example.hpp:3]: `example.hpp` or `indentity.hpp`? Same for `.cpp` files.

File: include/Beman/Example/example.hpp
Comment (neatudarius) [inline include/Beman/Example/example.hpp:35]: Are we OK with this?

Note: `std::forward` and `rvalue` references available since C++11; we need another implementation for C++98. Not sure if would bring any value.

File: .github/CODEOWNERS
Comment (camio) [reply .github/CODEOWNERS:4]: I think the error message is resolved now, right?

File: .gitignore
Comment (camio) [inline .gitignore:8]:
```suggestion
```
Editor specific files belong in a user's global .gitignore file. I think this should list only those files that this repository will create as part of its build process. That'd be only the build directory, right?

File: CMakeLists.txt
Comment (camio) [inline CMakeLists.txt:1]:
```suggestion
```
I'm unconvinced these file identifier lines are standard practice or worthwhile.

Comment (steve-downey) [reply CMakeLists.txt:1]: Depends where you've been in industry. Having had to do code inspection reviews with things printed out means that things need self identification. There is no IDE. 

Tagging the kind of file is fairly common 1.4M hits for  "-*- C++ -*-" https://github.com/search?q=%22-*-+c%2B%2B+-*-%22&type=code  65K for "-*- Makefile -*-" 

It of course shows up in all FSF software, since emacs uses it to identify the mode for the file. Definitely an issue when you have C++ in dot h files. 

There's more divergence in file name, with in addition to just the bare name here, using \file, @file, or other markup for identifying the work under copyright. The name was originally the canonical name for the file, not a position in the repository. For headers this is useful because it documents what the spelling of the include should be. That is, it should always be spelled `#include <Beman/Optional26/optional.hpp>` not `optional.hpp` or some other form. Implementation files don't really need a canonical name, particularly if the library is not spread across multiple directories, and the CML.txt is local to the file being added to the sources list for a file. 

Yes, this particular form comes from the BDE style guide, which originated with Mentor Graphics ages ago. I've got a yasnippet macro that fills out a header and implementation when I create a new file. The guide is of course highly pedantic and detail oriented, but at least it's all written down: 
https://bloomberg.github.io/bde/knowledge_base/coding_standards.html#prologue

File: .github/CODEOWNERS
Comment (neatudarius) [reply .github/CODEOWNERS:4]: I think bbrown105 is not added, but bretbrownjr is :D

File: .gitignore
Comment (neatudarius) [reply .gitignore:8]: agree. please check again

File: CMakeLists.txt
Comment (neatudarius) [reply CMakeLists.txt:1]: I will put this on our agenda. I would like to get a consensus, and make a recommendation.

File: src/Beman/Example/CMakeLists.txt
Comment (camio) [inline src/Beman/Example/CMakeLists.txt:14]: We should also be adding the headers here (as part of a FILE_SET) so they show up as part of the beman_example target in IDEs and are installed automatically as part of `make install`. I see something like that below, but it is commented out.

File: src/Beman/Example/CMakeLists.txt
Comment (camio) [inline src/Beman/Example/CMakeLists.txt:25]: I think we should be using FILE_SET but here it is commented out. Would it also remove the need for the `include_directories` call?

File: src/Beman/Example/CMakeLists.txt
Comment (camio) [inline src/Beman/Example/CMakeLists.txt:51]: IMO the tests should live right next to the source files.

File: src/Beman/Example/CMakeLists.txt
Comment (camio) [inline src/Beman/Example/CMakeLists.txt:44]: Is there a way to do this without including GNUInstalDirs?

File: src/Beman/Example/example.cpp
Comment (camio) [inline src/Beman/Example/example.cpp:4]:
```suggestion
```

File: src/Beman/Example/tests/CMakeLists.txt
Comment (camio) [inline src/Beman/Example/tests/CMakeLists.txt:7]:
```suggestion
add_executable(Beman.Example.test)
```

File: src/Beman/Example/tests/CMakeLists.txt
Comment (camio) [inline src/Beman/Example/tests/CMakeLists.txt:10]:
```suggestion
target_sources(beman_example_test
```
Omit needless words

File: src/Beman/Example/tests/example.t.cpp
Comment (camio) [inline src/Beman/Example/tests/example.t.cpp:16]: I suggest putting all includes at the top and making them unconditional. That simplifies this code.

File: CMakeLists.txt
Comment (neatudarius) [reply CMakeLists.txt]: Removed after our sync: `Allow comments for tooling, not recommended or required by the Beman Standard. (Outcome: Unanimous consent)`

File: README.md
Comment (neatudarius) [reply README.md:7]: Please check Optional26 (https://github.com/beman-project/Optional26). I think it's useful to see the latest status of main, and I think we may have multiple CI jobs for other projects - e.g., Unit tests, integration tests, or Linux/Windows tests.


image

File: README.md
Comment (neatudarius) [reply README.md:23]: OK., I will temporary remove it, but I created a TODO in PR description to add a OS support section.

File: README.md
Comment (neatudarius) [reply README.md:194]: I think still we need to specify the standard support. Temporary remove it.

File: src/Beman/Example/CMakeLists.txt
Comment (neatudarius) [reply src/Beman/Example/CMakeLists.txt:28]: I would like to first make a recommendation for the Beman Standard. On my topics for sync agenda.

File: src/Beman/Example/tests/CMakeLists.txt
Comment (neatudarius) [reply src/Beman/Example/tests/CMakeLists.txt]: I seems that cannot compile without it.

File: src/Beman/Example/CMakeLists.txt
Comment (neatudarius) [reply src/Beman/Example/CMakeLists.txt]: removed

File: README.md
Comment (neatudarius) [reply README.md]: Added a TODO, I will solved it in this PR.

File: README.md
Comment (neatudarius) [reply README.md]: removed

File: README.md
Comment (neatudarius) [reply README.md]: done

File: README.md
Comment (neatudarius) [reply README.md]: Agree. Created https://github.com/beman-project/Example/issues/17.

File: .github/CODEOWNERS
Comment (neatudarius) [inline .github/CODEOWNERS:4]:
```suggestion
*  @bretbrownjr @camio @dietmarkuehl  @neatudarius @steve-downey 
```

File: .github/CODEOWNERS
Comment (neatudarius) [inline .github/CODEOWNERS:4]:
```suggestion
*  @bretbrownjr @camio @dietmarkuehl @neatudarius @steve-downey 
```

File: src/Beman/Example/CMakeLists.txt
Comment (camio) [inline src/Beman/Example/CMakeLists.txt:8]:
```suggestion
add_library(beman.example STATIC)
add_library(beman::example ALIAS beman.example)
```

IMO it is much simpler to avoid defining variables like `TARGET_LIBRARY` that simply expand to another variable. It's better to use that variable directly.

While I appreciate the intent to avoid repetition by setting the library name to the project name, I think this aspect would be better handled by a linter than introducing this complexity.

File: src/Beman/Example/identity.cpp
Comment (camio) [inline src/Beman/Example/identity.cpp:3]:
```suggestion
#include <beman/example/identity.hpp>
```

File: src/Beman/Example/tests/CMakeLists.txt
Comment (camio) [inline src/Beman/Example/tests/CMakeLists.txt:7]:
```suggestion
add_executable(beman.example.tests.identity)
```
Again, using variable names makes the code less obscure.

Comment (camio) [inline src/Beman/Example/tests/CMakeLists.txt:7]: Should we follow the same pattern used in examples here?

File: src/Beman/Example/tests/identity.t.cpp
Comment (camio) [inline src/Beman/Example/tests/identity.t.cpp:3]:
```suggestion
#include <beman/example/identity.hpp>
```

File: examples/identity_usage.cpp
Comment (camio) [inline examples/identity_usage.cpp:3]:
```suggestion
#include <beman/example/identity.hpp>
```

File: examples/identity_usage.cpp
Comment (camio) [inline examples/identity_usage.cpp:8]:
```suggestion
    std::cout << beman::example::identity()(2024) << '\n';
```

File: src/Beman/Example/tests/CMakeLists.txt
Comment (camio) [inline src/Beman/Example/tests/CMakeLists.txt:9]: Since we are already in a `test/` subdirectory, I don't think it is necessary to use `.t` in the filename.

**Review by camio — COMMENTED** [review]
I wasn't able to add suggestions for the file and directory names, but those should be updated as well.

File: README.md
Comment (neatudarius) [reply README.md]: Defer to other PR. Agree with you.

File: src/Beman/Example/tests/CMakeLists.txt
Comment (neatudarius) [reply src/Beman/Example/tests/CMakeLists.txt]: Moved tests near the production code.

File: examples/CMakeLists.txt
Comment (camio) [reply examples/CMakeLists.txt]: I assume this is agreement with what I said and that you think it can come in another PR.

File: src/beman/example/CMakeLists.txt
Comment (camio) [inline src/beman/example/CMakeLists.txt:26]: Should the target name be `beman.example.tests.identity` in case there are multiple test drivers?

**Review by camio — APPROVED** [review]
Aside from the TODOs, which can easily be converted into issues against this repository, and a minor issue, this looks good to merge to me.

File: .github/workflows/ci_tests.yml
Comment (wusatosi) [inline .github/workflows/ci_tests.yml:37]: > Extend compiler / standard versions for build and run on CI: {gcc, clang} x {C++98, C++03, C++11, C++14, C++17, C++20, 'C++23, C++26}.

Is removing CI testing for C++98 <-> C++17 here intentional?
There's only C++20 <-> C++26 testing here.

@neatudarius

**Pull request merged**
